File: Final Project/code/metropolis.py
import numpy as np
import physics

def propose_phi(phi: np.ndarray,
                width: float,
                rng: np.random.Generator) -> np.ndarray:
    delta = 2. * rng.random(size=phi.shape) - 1.  # 2D array with random numbers between -1 and 1.
    return phi + width * delta


def metropolis_step(phi: np.ndarray,
                    mass2: float,
                    lamb: float,
                    width: float,
                    rng: np.random.Generator):
    # 1. Proposal
    proposed_phi = propose_phi(phi, width, rng)

    # 2. Acceptance probability.
    current_action = physics.phi4_action(phi, mass2=mass2, lamb=lamb)
    proposed_action = physics.phi4_action(proposed_phi, mass2=mass2, lamb=lamb)
    p_acceptance = np.min([1.0, np.exp(current_action - proposed_action)])

    # 3. Accept / reject
    r = rng.random()

    if r <= p_acceptance:  # Accept.
        return proposed_phi, 1

    # Reject.
    return phi, 0


def sample_field(initial_phi: np.ndarray,
                 num_samples: int,
                 mass2: float,
                 lamb: float,
                 width: float,
                 rng: np.random.Generator):
    chain = [initial_phi]
    acceptance = []

    for _ in range(num_samples):
        phi, accepted = metropolis_step(phi=chain[-1], mass2=mass2, lamb=lamb, width=width, rng=rng)
        chain.append(phi)
        acceptance.append(accepted)

    return chain, acceptance

File: Final Project/code/rust/src/metropolis.rs
use rand::Rng;

pub struct MetropolisSim<R>
where 
    R: Rng, 
{
    pub phi: Vec<f64>, //no longer a np.array

    pub mass2: f64, 
    pub lamb: f64, 
    pub eps: f64, 

    pub rng: R, 

    pub accepted_history: Vec<u8>, 
    pub integrator: I, 
    pub steps: usize, 
    pub bc: BoundaryCondition, 
}

impl<R> BaseSim for MetropolisSim<R>
where
    R: Rng, 
{
    pub type ActionFn = fn(phi: &[f64], mass2: f64, lamb: f64, bc: BoundaryCondition) -> f64;
    
    pub fn propose_phi(&mut self) -> Vec<f64> {
        self.phi.iter().map(|&x|{
            let delta: f64 = 2.0*self.rng.gen::<f64>() - 1.0; 
            x + self.width*delta
        }).collect()
    }

    fn update(&mut self) {
        let proposed_phi = self.propose_phi();

        let current_action =
            (self.action)(&self.phi, self.mass2, self.lamb, self.bc);
        let proposed_action =
            (self.action)(&proposed_phi, self.mass2, self.lamb, self.bc);

        let p_accept = (current_action - proposed_action)
            .exp()
            .min(1.0);

        let r: f64 = self.rng.gen();// random number in [0,1)

        if r <= p_accept {
            self.phi = proposed_phi;
            self.accepted_history.push(1);
        } else {
            self.accepted_history.push(0);
        }
    }

    fn get_steps(&self) -> usize {
        self.accepted_history.len()
    }
}

pub fn run_metropolis<R>(
    phi0: Vec<f64>,
    mass2: f64,
    lamb: f64,
    width: f64,
    n_steps: usize,
    rng: R,
    action: ActionFn,
    logger: Option<impl FnMut(&MetropolisSim<R>)>,
    bc: BoundaryCondition,
) -> Vec<u8>
where
    R: Rng,
{
    let mut sim = MetropolisSim {
        phi: phi0,
        mass2,
        lamb,
        width,
        rng,
        accepted_history: Vec::new(),
        bc,
        action,
    };

    sim.run_sim(n_steps, logger);

    sim.accepted_history
}
